File: platform/leetcode/39/39.py
from typing import List


def combinationSum(candidates: List[int], target: int) -> List[List[int]]:
    candidates = sorted(candidates)

    def dfs(start, remaining):
        if remaining == 0:
            return [[]]
        else:
            res = []
            for i in range(start, len(candidates)):
                c = candidates[i]
                if c > remaining:
                    break
                for rest in dfs(i, remaining - c):
                    res.append([c] + rest)
            return res

    return dfs(0, target)

File: platform/leetcode/39/39.hs
combinationSum :: [Int] -> Int -> [[Int]]
combinationSum _ 0 = [[]]
combinationSum [] _ = []
combinationSum (c:cs) target
  | target < 0 = []
  | otherwise = map (c:) (combinationSum (c:cs) (target-c)) ++ combinationSum cs target

main = print (combinationSum [2,3,6,7] 7)
